torch/csrc/jit/codegen/cuda/expr_simplifier.h

```diff
@@ -0,0 +1,533 @@
+// Proof: Same proof as 4.14
+//
+// Theorem 6.15: If compatible_sign(a, b) and -|c| < a % c + b % c < |c|, then
+// (a+b)/c = a/c + b/c
+// Proof: From Theorem 6.7 and Theorem 6.5
+// (a + b) % c = (a % c + b % c) % c = a % c + b % c
+// The rest of the proof is the same as 4.15
+//
+// 7. Summary
+// ==========
+//
+// In summary, in nvFuser, we can use the following rules to simplify integer
+// expressions:
+//
+// A) Associativity of +: a + (b + c) = (a + b) + c
+// B) Associativity of *: a * (b * c) = (a * b) * c
+// C) Commutativity of +: a + b = b + a
+// D) Commutativity of *: a * b = b * a
+// E) Distributivity of * over +: a * (b + c) = (a * b) + (a * c)
+// F) Distributivity of * over +: (a + b) * c = (a * c) + (b * c)
+// G) (-a) / b = -(a / b) = a / (-b)
+// H) (-a) % b = -(a % b) = a % (-b)
+// I) If -|a| < r < |a|, then r % a = r, r / a = 0
+// J) Distributivity of % over +:
+//    If compatible_sign(a, b), then (a + b) % c = (a % c + b % c) % c
+// K) Distributivity of % over *:
+//    If compatible_sign(a, b), then (a * b) % c = (a % c * b % c) % c
+// L) If a is a multiple of b, then a % b = 0
+// M) If b is a multiple of c, then we have: a*(b/c) = (a*b)/c
+// N) a / (b * c) = (a / b) / c
+// O) If d divides a and b, then a % b = ((a / d) % (b / d)) * d
+// P) If b is a multiple of c, then a/(b/c) = (a*c)/b
+// Q) If compatible_sign(a, b) and -|c| < a % c + b % c < |c|, then
+//    (a+b)/c = a/c + b/c
+//
+// Some examples on applying the above rules to simplify expressions:
+//
+// Example 7.1: Given that a >= 0 and b >= 0, simplify (a*4 + b) % 4
+// Answer: (a*4 + b) % 4 = ((a*4)%4 + b%4) % 4 (Rule J)
+// = (0 + b%4) % 4 (Rule L)
+// = b % 4 % 4 (Basic math)
+// = b % 4 (Rule I)
+//
+// Example 7.2: Given that 0 <= a < 3, simplify a % 4
+// Answer: a % 4 = a (Rule I)
+//
+// Example 7.3: Simplify (a * 256) / 4
+// Answer: (a * 256) / 4 = a * (256 / 4) (Rule M)
+// = a * 64 (Basic math)
+//
+// Example 7.4: Simplify (a / 4) / 64
+// Answer: (a / 4) / 64 = a / (4 * 64) (Rule N)
+// = a / 256 (Basic math)
+//
+// Example 7.5: Simplify (a * 64) % 256 / 4
+// Answer: (a * 64) % 256 / 4 = ((a % 4) * 64) / 4 (Rule O)
+// = (a % 4) * (64 / 4) (Rule M)
+// = (a % 4) * 16 (Basic math)
+//
+// Example 7.6: Simplify (a * 4) / 256
+// Answer: (a * 4) / 256 = a / (256 / 4) (Rule P)
+// = a / 64 (Basic math)
+//
+// Example 7.7: Given that a >= 0 and b >= 0, simplify (a * 256 + b) / 4
+// Answer: because (a * 256) % 4 = 0, we have
+// (a * 256 + b) / 4 = a * 256 / 4 + b / 4 (Rule Q)
+// = a * (256 / 4) + b / 4 (Rule M)
+// = a * 64 + b / 4 (Basic math)
+//
+// Example 7.8: Given that a >= 0 and 0 <= b < 4, simplify (a * 4 + b) / 4
+// Answer: Similar to above, we have
+// (a * 4 + b) / 4 = a + b / 4
+// = a + 0 (Rule I)
+// = a
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+struct ValInfo {
+  Val *lower_bound; // including
+  Val *upper_bound; // TODO: excluding?
```

Comment:
I assume this is trying to provide some relative properties between integers?

Reply:
It tries to keep track of the range of loop index. For example, in `FOR i in 0...N`, we have `lower_bound = 0` and `upper_bound = N`, then we can use this information to prove `i >= 0` and `i < N` during simplification.

Reply:
gotcha, I wonder what other properties we might want to extract out of an IterDomain pass.

Reply:
`step` might also be helpful. If a loop variable `i1` has `start == 0` and `step == 2`, then we can prove that `i1 % 2 == 0`